core/src/test/scala/unit/kafka/server/MetadataRequestTest.scala

```diff
@@ -0,0 +1,156 @@
+    val props = TestUtils.createBrokerConfigs(numBrokers, zkConnect, enableControlledShutdown = false)
+    props.foreach { p =>
+      p.setProperty(KafkaConfig.DeleteTopicEnableProp, "true")
+      p.setProperty(KafkaConfig.RackProp, s"rack/${p.getProperty(KafkaConfig.BrokerIdProp)}")
+    }
+    props.map(KafkaConfig.fromProps)
+  }
+
+  private def allMetadataRequest = new MetadataRequest(null, true)
+  private def noMetadataRequest = new MetadataRequest(List[String]().asJava)
+
+  @Before
+  override def setUp() {
+    super.setUp()
+  }
+
+  @Test
+  def testControllerId() {
+    val metadataResponse = sendMetadataRequest(allMetadataRequest, 1)
+    val controllerServer = servers.find(_.kafkaController.isActive()).get
+    val controllerId = controllerServer.apis.brokerId
+
+    assertEquals("Controller id should match the active controller",
+      controllerServer.apis.brokerId, controllerId)
+
+    // Fail over the controller
+    controllerServer.shutdown()
+    controllerServer.startup()
+
+    val metadataResponse2 = sendMetadataRequest(allMetadataRequest, 1)
+    val controllerServer2 = servers.find(_.kafkaController.isActive()).get
+    val controllerId2 = controllerServer2.apis.brokerId
+
+    assertNotEquals("Controller id should switch to a new broker", controllerId, controllerId2)
+    assertEquals("Controller id should match the active controller after failover",
+      controllerServer.apis.brokerId, metadataResponse.controller.id)
+  }
+
+  @Test
+  def testRack() {
+    val metadataResponse = sendMetadataRequest(allMetadataRequest, 1)
+    // Validate rack matches whats set in generateConfigs() above
+    metadataResponse.brokers().asScala.foreach { broker =>
+      assertEquals("Rack information should match config", s"rack/${broker.id}", broker.rack())
+    }
+  }
+
+  @Test
+  def testIsInternal() {
+    val internalTopic = TopicConstants.GROUP_METADATA_TOPIC_NAME
+    val notInternalTopic = "notInternal"
+    // create the topics
+    TestUtils.createTopic(zkUtils, internalTopic, 3, 2, servers)
+    TestUtils.createTopic(zkUtils, notInternalTopic, 3, 2, servers)
+
+    val metadataResponse = sendMetadataRequest(allMetadataRequest, 1)
+    assertTrue("Response should have no errors", metadataResponse.errors().isEmpty)
+
+    val topicMetadata = metadataResponse.topicMetadata().asScala
+    val internalTopicMetadata = topicMetadata.find(_.topic == internalTopic).get
+    val notInternalTopicMetadata = topicMetadata.find(_.topic == notInternalTopic).get
+
+    assertTrue("internalTopic should show isInternal", internalTopicMetadata.isInternal)
+    assertFalse("notInternalTopic topic not should show isInternal", notInternalTopicMetadata.isInternal)
+  }
+
+  @Test
+  def testNoTopicsRequest() {
+    // create some topics
+    TestUtils.createTopic(zkUtils, "t1", 3, 2, servers)
+    TestUtils.createTopic(zkUtils, "t2", 3, 2, servers)
+
+    val metadataResponse = sendMetadataRequest(noMetadataRequest, 1)
```

Comment:
Do you want to also validate that V0 request with null still returns all topics?